File: 计算机工程实践/Transbot SE/pi/src/catkin_ws/src/learning_topic/scripts/Information_subscriber.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


import rospy

from learning_topic.msg import Information #导入自定义的msg

def CompanyInfoCallback(msg):

    rospy.loginfo("company: name:%s  city:%s ", msg.company, msg.city)#打印订阅接收到信息

def Infomation_subscriber():
    
    rospy.init_node('Infomation_subscriber', anonymous=True)# ROS节点初始化

    # 创建一个Subscriber，订阅名为/company_info的topic，注册回调函数personInfoCallback
    rospy.Subscriber("/company_info", Information, CompanyInfoCallback)

    rospy.spin()# 循环等待回调函数

if __name__ == '__main__':
    Infomation_subscriber()

File: 计算机工程实践/Transbot SE/pi/src/catkin_ws/src/learning_topic/src/Information_subscriber.cpp
/**
 * 该例程将订阅/company_info话题，自定义消息类型learning_topic::Information
 */
 
#include <ros/ros.h>
#include "learning_topic/Information.h"

// 接收到订阅的消息后，会进入消息回调函数处理数据
void CompanyInfoCallback(const learning_topic::Information::ConstPtr& msg)
{
    // 打印接收到的消息
    ROS_INFO("This is: %s  in %s", msg->company.c_str(), msg->city.c_str());
}

int main(int argc, char **argv)
{
    
    ros::init(argc, argv, "company_Information_subscriber");// 初始化ROS节点

    
    ros::NodeHandle n;// 这里是创建节点句柄

    // 创建一个Subscriber，订阅名话题/company_info的topic，注册回调函数CompanyInfoCallback
    ros::Subscriber person_info_sub = n.subscribe("/company_info", 10, CompanyInfoCallback);

    
    ros::spin();// 循环等待回调函数

    return 0;
}
